crates/validator-transaction-pool/src/lib.rs

```diff
@@ -27,104 +22,125 @@ impl TransactionFilter {
+pub struct PoolState {
+    /// Incremented every time a txn is pushed in. The txn gets the old value as its sequence number.
+    next_seq_num: u64,
 
-    /// Currently only support 1 txn per topic.
-    txns: HashMap<Topic, Arc<ValidatorTransaction>>,
+    /// Track Topic -> seq_num mapping.
+    /// We allow only 1 txn per topic and this index helps find the old txn when adding a new one for the same topic.
+    seq_nums_by_topic: HashMap<Topic, u64>,
 
-    pull_notification_senders: HashMap<Topic, Mutex<PullNotificationSender>>,
+    /// Txns ordered by their sequence numbers (i.e. time they entered the pool).
+    txn_queue: BTreeMap<u64, PoolItem>,
+}
+
+/// Returned for `txn` when you call `PoolState::put(txn, ...)`.
+/// If this is dropped, `txn` will be deleted from the pool (if it has not been).
+///
+/// This allows the pool to be emptied on epoch boundaries.
+pub struct TxnGuard {
+    pool: Arc<Mutex<PoolState>>,
+    seq_num: u64,
 }
 
 impl PoolState {
-    pub fn new(topic_sender_pairs: Vec<(Topic, Option<PullNotificationSender>)>) -> Self {
-        let topics: Vec<Topic> = topic_sender_pairs.iter().map(|(topic, _)| *topic).collect();
-        let topic_set: HashSet<Topic> = HashSet::from_iter(topics.clone());
-        assert_eq!(topics.len(), topic_set.len());
-        let pull_notification_senders = topic_sender_pairs
-            .into_iter()
-            .filter_map(|(topic, maybe_sender)| {
-                maybe_sender.map(|sender| (topic, Mutex::new(sender)))
-            })
-            .collect();
-        Self {
-            txns: HashMap::new(),
-            pull_notification_senders,
-            topics,
+    /// Append a txn to the pool.
+    /// Return a txn guard that allows you to later delete the txn from the pool.
+    pub fn put(
+        pool: Arc<Mutex<Self>>,
+        topic: Topic,
+        txn: Arc<ValidatorTransaction>,
+        pull_notification_tx: Option<aptos_channel::Sender<(), Arc<ValidatorTransaction>>>,
+    ) -> TxnGuard {
+        let mut pool_guard = pool.lock();
+        let seq_num = pool_guard.next_seq_num;
+        pool_guard.next_seq_num += 1;
+
+        pool_guard.txn_queue.insert(seq_num, PoolItem {
+            seq_num,
+            topic: topic.clone(),
+            txn,
+            pull_notification_tx,
+        });
+
+        if let Some(old_seq_num) = pool_guard.seq_nums_by_topic.insert(topic.clone(), seq_num) {
+            pool_guard.txn_queue.remove(&old_seq_num);
         }
-    }
-}
 
-impl Default for PoolState {
-    fn default() -> Self {
-        Self::new(vec![])
+        TxnGuard {
+            pool: pool.clone(),
+            seq_num,
+        }
     }
-}
 
-pub struct ReadClient {
-    pool: Arc<Mutex<PoolState>>,
-}
+    fn try_delete(&mut self, seq_num: u64) {
+        if let Some(item) = self.txn_queue.remove(&seq_num) {
+            let seq_num_another = self.seq_nums_by_topic.remove(&item.topic);
+            assert_eq!(Some(seq_num), seq_num_another);
```

Comment:
What is the invariant to ensure this assert? In `txn_queue` only one txn is kept for each topic, correct?

Reply:
Correct